### PR TITLE
Performing Vocabulary Parallelism for LM Head across Attention TP Groups

### PR DESCRIPTION
**Pull request by ch-wan**
## Motivation

This PR allows users to set `--enable-dp-lm-head` for accelerating `lm_head` for DP attention. The original implementation adopts vocabulary parallel across the entire TP group, which has two issues under DP attention:

  1. The memory requirement is proportional to the DP size, which may prevent the execution of CUDA Graph.
  2. Redundant communication is required. A scatter operation is called immediately after all-gather. See [here](https://github.com/sgl-project/sglang/blob/97cb762bb65ebf05025eb342de03c184660427a3/python/sglang/srt/layers/logits_processor.py#L444-L456).

## Modifications

We adopt vocabulary parallel across the *attention TP group* to avoid all-gather across DP groups. 

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by xu-yfei** [comment]
It seems that lm_head does dp, which consumes extra weight memory？it may be not very friendly to H20 (96GB). We plan to do dp on 8xH20 (96GB) #5001 , lm_head memory occupies 1.726 GB, and the 1.726*7/8 GB less will have a big impact on us

**Comment by ch-wan** [comment]
@xu-yfei Thank you for sharing your opinion. Could you please double confirm the memory consumption of the LM head? The vocabulary size and hidden dimension of DeepSeek are 129280 and 7168, respectively, which occupies 1.726GB memory **in total** in BF16.

**Comment by ch-wan** [comment]
@xu-yfei Sure, we can expose an interface to our user to decide whether to use vocabulary parallelism or data parallelism for LM head. It's worth noting that vocabulary parallelism does not necessarily save memory in this case as it requires to gather tokens across the entire TP group.

File: python/sglang/srt/layers/logits_processor.py [hunk 1]
```diff
@@ -198,12 +199,19 @@ def __init__(
...
-            not skip_all_gather and get_tensor_model_parallel_world_size() > 1
-        )
-        self.do_tensor_parallel_all_gather_dp_attn = (
-            self.do_tensor_parallel_all_gather and get_attention_dp_size() != 1
-        )
+        self.use_attn_tp_group = global_server_args_dict["enable_dp_lm_head"]
+        if self.use_attn_tp_group:
+            self.attn_tp_size = get_attention_tp_size()
+            self.do_tensor_parallel_all_gather = (
+                not skip_all_gather and self.attn_tp_size > 1
```
Comment (ch-wan) [inline python/sglang/srt/layers/logits_processor.py:206]: why do we have to do all-gather across the entire TP group when use attn_tp_group?

**Comment by KSLee7** [comment]
Hi, I tried `--enable-dp-lm-head` but effects are not obvious on DeepSeek-R1, with 2 nodes (8*H20 for each) and DP=2/4/8. 
`python -m sglang.launch_server --model-path /home/model/DeepSeek-R1 --dist-init-addr 127.0.0.1 --nnodes 2 --node-rank 0 --trust-remote-code --attention-backend fa3 --tp 16 --cuda-graph-max-bs 40 --enable-dp-attention --dp 4 --enable-dp-lm-head`

**Comment by ch-wan** [comment]
@KSLee7 Thank you for your benchmark. LM head is not the major efficiency bottleneck. We use `--enable-dp-lm-head` to reduce the batch size of CUDA Graph.

**Comment by KSLee7** [comment]
> @KSLee7 Thank you for your benchmark. LM head is not the major efficiency bottleneck. We use `--enable-dp-lm-head` to reduce the batch size of CUDA Graph.

Thx.

**Comment by bird2426** [comment]
Have you tried testing with the configuration from this blog (https://lmsys.org/blog/2025-05-05-large-scale-ep/)? Would it yield significant improvements in throughput and other performance metrics?

**Comment by ch-wan** [comment]
@lianzhanbiao This is the PR for implementing DP head in this blog.
